### PR TITLE
Multiple Institutions (Affiliated Institutions for Nodes)

### PR DESCRIPTION
**Pull request by hmoco**
## Purpose:
To allow users to affiliated a node with multiple institutions.

## Changes:
Removal of primary institution, and any functionality that dealt with primary institutions now talks to the affiliated institutions field in node instead. Settings page now supports affiliating multiple institutions, and so does the branding in the nodes main page (properly supports up to 4 institutions, with their logo showing up next to each other).
 
## Side-Effects:
Removes primary_institution field, and all api components associated with it. Requires a migration, which changes all the primary institution logs to affiliated institution logs. 

## Ticket:
https://openscience.atlassian.net/browse/OSF-5845

### CODE REVIEW

File: api/institutions/views.py [hunk 1]
```diff
@@ -8,18 +9,24 @@
...
+)
+from api.base.exceptions import RelationshipPostMakesNoChanges
 from api.nodes.serializers import NodeSerializer
 from api.users.serializers import UserSerializer
 
 from .authentication import InstitutionAuthentication
-from .serializers import InstitutionSerializer
-
+from .serializers import InstitutionSerializer, InstitutionNodesRelationshipSerializer
+from .permissions import UserIsAffiliated
```
Comment (mfraezz) [inline api/institutions/views.py]: I know we don't always do it, but generally prefer absolute over relative module paths

File: api/institutions/serializers.py [hunk 2]
```diff
@@ -40,3 +47,49 @@ def get_absolute_url(self, obj):
...
+        nodes = [Node.load(node_dict['_id']) for node_dict in node_dicts]
+
+        changes_flag = False
+        for node in nodes:
+            if not node:
+                raise exceptions.NotFound
+            if not node.has_permission(user, osf_permissions.ADMIN):
+                raise exceptions.PermissionDenied
+
+        for node in nodes:
```
Comment (mfraezz) [inline api/institutions/serializers.py]: Why not just do this in the above `for node in nodes` loop?

Comment (hmoco) [inline api/institutions/serializers.py]: Because wanted behavior is to not update any node if the request has any problems with it.

Comment (mfraezz) [inline api/institutions/serializers.py]: There are faster ways of accomplishing that, but it's not very important. I doubt we'll get too many requests here with a large number of nodes.

File: api/nodes/views.py [hunk 3]
```diff
@@ -1930,79 +1935,104 @@ class NodeInstitutionDetail(JSONAPIBaseView, generics.RetrieveAPIView, NodeMixin
...
-        inst = request.data.get('id', None)
-        if not inst:
+    def perform_destroy(self, instance):
+        data = self.request.data['data']
+        user = self.request.user
+        current_insts = {inst._id: inst for inst in instance['data']}
+        node = instance['self']
+        for val in data:
+            if val['id'] in current_insts:
+                node.remove_affiliated_institution(inst=current_insts[val['id']], user=user)
```
Comment (mfraezz) [inline api/nodes/views.py]: Unless we want to be quiet about failing to delete things that aren't there (or that don't exist), it seems like this method could use more error handling.

File: api_tests/institutions/views/test_institution_relationship_nodes.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+        )
+
+        assert_equal(res.status_code, 201)
+        node_ids = [node_['id'] for node_ in res.json['data']]
+        assert_in(node._id, node_ids)
+
+        node.reload()
+        assert_in(self.institution, node.affiliated_institutions)
+
+    def test_user_doesnt_have_node(self):
```
Comment (mfraezz) [inline api_tests/institutions/views/test_institution_relationship_nodes.py]: Trivial, but contractions shouldn't be used in method names

File: api_tests/nodes/views/test_node_relationship_institutions.py [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+        self.node.save()
+
+        res = self.app.get(
+            self.node_institutions_url
+        )
+
+        assert_equal(res.status_code, 200)
+        assert_equal(res.json['data'], [])
+
+    def test_institution_doesnt_exist(self):
```
Comment (mfraezz) [inline api_tests/nodes/views/test_node_relationship_institutions.py]: Here [also](https://github.com/CenterForOpenScience/osf.io/pull/5450/files#r62365983)

File: scripts/migration/migrate_primary_institutions.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    for node in nodes:
+        logger.warn('Node with id <{}> and title <{}> being updated'.format(node._id, node.title))
+        if not dry_run:
+            inst = node.primary_institution
+            if inst not in node.affiliated_institutions:
+                node.affiliated_institutions.append(inst)
+            node.primary_institution = None
+            node.save()
+
+def main(dry_run=True):
```
Comment (mfraezz) [inline scripts/migration/migrate_primary_institutions.py]: You need to add a file logger if it's not a `dry_run`. Also, the migration itself should be run within a `TokuTransaction`. This prevents changes from persisting if an exception is thrown during the migration, and during a `dry_run` allows you to actually run the migration to ensure it works.

```python
import logging
import sys

from framework.transactions.context import TokuTransaction
from website.app import init_app
from scripts import utils as script_utils

logger = logging.getLogger(__name__)

def migrate(dry_run=True):
    # iterate over targets
    # perform actual migration
    # log things

    if dry_run:
        raise RuntimeError('Dry run, transaction rolled back.')

def main():
    dry_run = False
    if '--dry' in sys.argv:
        dry_run = True
    if not dry_run:
        script_utils.add_file_logger(logger, __file__)
    init_app(set_backends=True, routes=False)
    with TokuTransaction():
        migrate(dry_run=dry_run)

if __name__ == "__main__":
    main()
```

File: api/institutions/views.py [hunk 7]
```diff
@@ -233,5 +240,89 @@ class InstitutionRegistrationList(InstitutionNodeList):
     def get_queryset(self):
         inst = self.get_institution()
         query = self.get_query_from_request()
-        nodes = list(Node.find_by_institution(inst, query))
+        nodes = list(Node.find_by_institutions(inst, query))
         return [node for node in nodes if not node.is_retracted]
```
Comment (mfraezz) [inline api/institutions/views.py]: This could be slow -- you're enumerating this queryset twice. ~~Why not just `return Node.find_by_institutions(inst, query & Q('is_retracted', 'eq', False))`?~~ Because `is_retracted` is a property. Still, you shouldn't need to `list` the queryset the first time.

**Comment by mfraezz** [comment]
Build is failing because `develop` is currently failing. Tests for this functionality look good:
```
DJANGO_SETTINGS_MODULE=api.base.settings nosetests api_tests/institutions/views/test_institution_relationship_nodes.py api_tests/institutions/views/test_institution_registrations_list.py api_tests/institutions/views/test_institution_nodes_list.py
-------------------------------------------------
Ran 24 tests in 17.148s
OK
```

A few minor issues to address and I'll do functional review. Pass complete 🛅

File: api/nodes/views.py (hunk 3 above)
Comment (hmoco) [inline api/nodes/views.py]: From json-api specs: 
```
If all of the specified resources are able to be removed from, or are already missing from, the relationship then the server MUST return a successful response.
```

Comment (mfraezz) [inline api/nodes/views.py]: There may be a distinction between `nonexistent` and `already missing from`.

File: api/nodes/views.py [hunk 8]
```diff
@@ -1930,79 +1935,104 @@ class NodeInstitutionDetail(JSONAPIBaseView, generics.RetrieveAPIView, NodeMixin
...
-        instance = self.get_object()
-        serializer = self.get_serializer(instance, data=request.data, partial=partial)
-        serializer.is_valid(raise_exception=True)
-        self.perform_update(serializer)
-        inst = request.data.get('id', None)
-        if not inst:
+    def perform_destroy(self, instance):
+        data = self.request.data['data']
+        user = self.request.user
+        current_insts = {inst._id: inst for inst in instance['data']}
```
Comment (mfraezz) [inline api/nodes/views.py]: It looks like this is the only place that `NodeInstitutionsRelationshipSerializer['data']` is being accessed; [other](https://github.com/CenterForOpenScience/osf.io/pull/5450/files#diff-7a6d21a50624153d15e63a5f643401c7R504) [places](https://github.com/CenterForOpenScience/osf.io/pull/5450/files#diff-7a6d21a50624153d15e63a5f643401c7R526) just use `instance['self'].affiliated_institutions`. Why can that not be done here, or more generally, why is wrapping a `Node` in a `dict(self=node, data=node.<attribute>)` necessary?

File: api/institutions/views.py [hunk 9]
```diff
@@ -29,12 +36,12 @@ class InstitutionMixin(object):
 
     def get_institution(self):
         inst = get_object_or_error(
-            Node,
-            Q('institution_id', 'eq', self.kwargs[self.institution_lookup_url_kwarg]),
+            Institution,
+            self.kwargs[self.institution_lookup_url_kwarg],
             display_name='institution',
             allow_institution=True
```
Comment (mfraezz) [inline api/institutions/views.py]: The `Institution.find()` override automatically includes `allow_institution` in a `Node.find` call; you should be able to remove this line.

File: scripts/migration/migrate_primary_institutions.py [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+logger = logging.getLogger(__name__)
+logging.basicConfig(level=logging.WARN)
+
+PRIMARY_INSTITUTION_CHANGED = 'primary_institution_changed'
+PRIMARY_INSTITUTION_REMOVED = 'primary_institution_removed'
+
+def migrate(dry_run=True):
+    added_logs = NodeLog.find(Q('action', 'eq', PRIMARY_INSTITUTION_CHANGED))
+    for log in added_logs:
+        logger.warn('Log with id <{}> being updated for affiliation added'.format(log._id))
```
Comment (mfraezz) [inline scripts/migration/migrate_primary_institutions.py]: Prefer `.info` for log messages that aren't indicative of something dangerous or failing.

File: scripts/migration/migrate_primary_institutions.py [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+        log.action = NodeLog.AFFILIATED_INSTITUTION_ADDED
+        log.save()
+
+    removed_logs = NodeLog.find(Q('action', 'eq', PRIMARY_INSTITUTION_REMOVED))
+    for log in removed_logs:
+        logger.warn('Log with id <{}> being updated for affiliation removed'.format(log._id))
```
Comment (mfraezz) [inline scripts/migration/migrate_primary_institutions.py]: `.info`

File: scripts/migration/migrate_primary_institutions.py [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+        log.action = NodeLog.AFFILIATED_INSTITUTION_REMOVED
+        log.save()
+
+    nodes = Node.find(Q('primary_institution', 'ne', None))
+    for node in nodes:
+        logger.warn('Node with id <{}> and title <{}> being updated'.format(node._id, node.title))
```
Comment (mfraezz) [inline scripts/migration/migrate_primary_institutions.py]: `.info`

File: scripts/migration/migrate_primary_institutions.py [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+from modularodm import Q
+
+from framework.transactions.context import TokuTransaction
+
+from scripts import utils
+from website.app import init_app
+from website.models import Node, NodeLog
+
+logger = logging.getLogger(__name__)
+logging.basicConfig(level=logging.WARN)
```
Comment (mfraezz) [inline scripts/migration/migrate_primary_institutions.py]: Why is this here? `INFO` level logs shouldn't be removed, and there isn't any error handling in this script that would merit a `WARN` log

File: scripts/migration/migrate_primary_institutions.py [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+        log.action = NodeLog.AFFILIATED_INSTITUTION_ADDED
+        log.save()
+
+    removed_logs = NodeLog.find(Q('action', 'eq', PRIMARY_INSTITUTION_REMOVED))
+    for log in removed_logs:
+        logger.warn('Log with id <{}> being updated for affiliation removed'.format(log._id))
+        log.action = NodeLog.AFFILIATED_INSTITUTION_REMOVED
+        log.save()
+
+    nodes = Node.find(Q('primary_institution', 'ne', None))
```
Comment (mfraezz) [inline scripts/migration/migrate_primary_institutions.py]: This might not work, as you're removing `primary_institution` from the model. Can still use pymongo, something like
```python
from framework.mongo import database

nodes = database['node'].find({'_primary_institution': {'$ne': None}})
```

File: scripts/migration/migrate_primary_institutions.py [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+    if dry_run:
+        raise RuntimeError('Dry run, transaction rolled back.')
+
+def main():
+    dry_run = 'dry' in sys.argv
+    if dry_run:
+        logger.warn('DRY RUN mode')
+    else:
+        utils.add_file_logger(logger, __file__)
+    init_app()
```
Comment (mfraezz) [inline scripts/migration/migrate_primary_institutions.py]:
```python
init_app(routes=False)
```

File: website/project/model.py [hunk 16]
```diff
@@ -3626,62 +3610,59 @@ def affiliated_institutions(self):
...
+                        }
+                    },
+                    auth=Auth(user)
+                )
+            if save:
+                self.save()
+            return True
+        return False
 
     def institution_url(self):
```
Comment (mfraezz) [inline website/project/model.py]: Are these valid any longer?
```
~/D/osf.io ❯ cat api/**/urls.py | ag institution/                 ⏎ remotes/origin/pr/5450 ✭
~/D/osf.io ❯ cat api/**/urls.py | ag institutions/                ⏎ remotes/origin/pr/5450 ✭
                url(r'^institutions/', include('api.institutions.urls', namespace='institutions')),
    url(r'^(?P<node_id>\w+)/institutions/$', views.NodeInstitutionsList.as_view(), name=views.NodeInstitutionsList.view_name),
    url(r'^(?P<node_id>\w+)/relationships/institutions/$', views.NodeInstitutionsRelationship.as_view(), name=views.NodeInstitutionsRelationship.view_name),
    url(r'^(?P<node_id>\w+)/institutions/$', views.RegistrationInstitutionsList.as_view(), name=views.RegistrationInstitutionsList.view_name),
    url(r'^(?P<user_id>\w+)/institutions/$', views.UserInstitutions.as_view(), name=views.UserInstitutions.view_name),
    url(r'^(?P<user_id>\w+)/relationships/institutions/$', views.UserInstitutionsRelationship.as_view(), name=views.UserInstitutionsRelationship.view_name),
```

File: website/search/elastic_search.py [hunk 17]
```diff
@@ -338,7 +338,7 @@ def update_node(node, index=None, bulk=False):
             'parent_id': parent_id,
             'date_created': node.date_created,
             'license': serialize_node_license_record(node.license),
-            'primary_institution': node.primary_institution.name if node.primary_institution else None,
+            'affiliated_institutions': [inst.name for inst in node.affiliated_institutions] if node.affiliated_institutions else [],
```
Comment (mfraezz) [inline website/search/elastic_search.py]: Minor, but if `node.affiliated_institutions` returns `[]`, `[inst.name for inst in node.affiliated_institutions]` will already be `[]`:

```
In [1]: [i.some_attr for i in []]
Out[1]: []
```

File: website/static/js/logActionsList.json [hunk 18]
```diff
@@ -65,6 +65,6 @@
     "osf_storage_folder_created" : "${user} created folder ${path} in OSF Storage in ${node}",
     "osf_storage_file_updated" : "${user} updated file ${path} in OSF Storage in ${node}",
     "osf_storage_file_removed" : "${user} removed file ${path} from OSF Storage in ${node}",
-    "primary_institution_changed": "${user} changed the primary institution in ${node} from ${previous_institution} to ${institution}.",
-    "primary_institution_removed": "${user} removed ${institution} as the primary institution in ${node}."
+    "affiliated_institution_added": "${user} added affiliation to ${institution} to ${node}.",
```
Comment (mfraezz) [inline website/static/js/logActionsList.json]: It seems like this would make more sense as `"${user} added ${institution} affiliation to ${node}"`

@saradbowman, thoughts?

File: website/templates/project/project.mako [hunk 19]
```diff
@@ -116,16 +118,23 @@
...
+                                % endif
+                            % endfor
                         % else:
                             <span> None </span>
                         % endif
                     % endif
-                    % if not ('admin' in user['permissions'] and not node['is_registration']) and node['institution']['id']:
-                        Affiliated institution: <a href="/institutions/${node['institution']['id']}">${node['institution']['name']}</a>
+                    % if not ('admin' in user['permissions'] and not node['is_registration']) and node['institutions'] != []:
+                        Affiliated institutions: <a href="/institutions/${node['institutions'][0]['id']}">${node['institutions'][0]['name']}</a>
```
Comment (mfraezz) [inline website/templates/project/project.mako]: Should this be iterating over `node['institutions']`?

**Comment by mfraezz** [comment]
A few issues, mostly small things. I'll test the migration script after the above comments on it are addressed. Also, please resolve merge conflicts. 

Pass complete 🍣

File: website/static/js/logActionsList.json (hunk 18 above)
Comment (saradbowman) [inline website/static/js/logActionsList.json]: I absolutely agree with @mfraezz.

File: website/templates/log_templates.mako [hunk 20]
```diff
@@ -421,15 +421,12 @@ from
...
+added affiliation to <a class="log-node-title-link overflow" data-bind="attr: {href: '/institutions/' + params.institution.id}, text: params.institution.name"></a>
+ to <a class="log-node-title-link overflow" data-bind="text: nodeTitle, attr: {href: nodeUrl}"></a>
 </script>
 
-<script type="text/html" id="primary_institution_removed">
-removed <a class="log-node-title-link overflow" data-bind="attr: {href: '/institutions/' + params.institution.id}, text: params.institution.name"></a>
-as the primary institution of <a class="log-node-title-link overflow" data-bind="text: nodeTitle, attr: {href: nodeUrl}"></a>.
+<script type="text/html" id="affiliated_institution_removed">
+removed affiliation to <a class="log-node-title-link overflow" data-bind="attr: {href: '/institutions/' + params.institution.id}, text: params.institution.name"></a>
+ in <a class="log-node-title-link overflow" data-bind="text: nodeTitle, attr: {href: nodeUrl}"></a>.
```
Comment (mfraezz) [inline website/templates/log_templates.mako]: Language change in this file too

File: api/nodes/views.py (hunk 8 above)
Comment (hmoco) [inline api/nodes/views.py]: This is a workaround to allow the serializer to properly put this data under the key ` data `.

**Comment by hmoco** [comment]
@mfraezz, responded to code review!

File: api/nodes/serializers.py [hunk 21]
```diff
@@ -485,58 +485,88 @@ def get_absolute_url(self, obj):
...
-                raise exceptions.NotFound
-            try:
-                node.add_primary_institution(inst=inst, user=user)
-            except UserNotAffiliatedError:
-                raise exceptions.ValidationError(detail='User not affiliated with institution')
-            node.save()
-            return node
-        node.remove_primary_institution(user)
+        add, remove = self.get_institutions_to_add_remove(
+            institutions=node.affiliated_institutions,
```
Comment (mfraezz) [inline api/nodes/serializers.py]: Equivalent to `instance['data']`, consider normalizing with [this comment](https://github.com/CenterForOpenScience/osf.io/pull/5450#discussion_r62914890)

File: api/nodes/serializers.py [hunk 22]
```diff
@@ -485,58 +485,88 @@ def get_absolute_url(self, obj):
...
+        return self.make_instance_obj(node)
 
-        return data
+    def create(self, validated_data):
+        instance = self.context['view'].get_object()
+        user = self.context['request'].user
+        node = instance['self']
+
+        add, remove = self.get_institutions_to_add_remove(
+            institutions=node.affiliated_institutions,
```
Comment (mfraezz) [inline api/nodes/serializers.py]: Equivalent to `instance['data']`, consider normalizing with [this comment](https://github.com/CenterForOpenScience/osf.io/pull/5450#discussion_r62914890)

File: api/nodes/views.py (hunk 8 above)
Comment (mfraezz) [inline api/nodes/views.py]: That's fine, but please be consistent in access methods (`instance['self'].affiliated_institutions` vs `instance['data']`).

File: api/institutions/serializers.py (hunk 2 above)
Comment (mfraezz) [inline api/institutions/serializers.py]: The API should be wrapping each request in a transaction, which would be rolled back on error. 

It looks like this is being tested in `api_tests/institutions/views/test_institution_relationship_nodes.py:TestInstitutionRelationshipNodes.test_add_some_with_permissions_others_without`, and when I modify this function to do both in the same loop, the test still passes.

File: api/institutions/serializers.py [hunk 23]
```diff
@@ -40,3 +47,49 @@ def get_absolute_url(self, obj):
...
+        user = self.context['request'].user
+        node_dicts = validated_data['data']
+        nodes = [Node.load(node_dict['_id']) for node_dict in node_dicts]
+
+        changes_flag = False
+        for node in nodes:
+            if not node:
+                raise exceptions.NotFound
+            if not node.has_permission(user, osf_permissions.ADMIN):
+                raise exceptions.PermissionDenied
```
Comment (mfraezz) [inline api/institutions/serializers.py]: Can you give this a descriptive error message? e.g. `"Admin permission on node <id> required."`

File: api/institutions/serializers.py [hunk 24]
```diff
@@ -40,3 +47,49 @@ def get_absolute_url(self, obj):
...
+    def create(self, validated_data):
+        inst = self.context['view'].get_object()['self']
+        user = self.context['request'].user
+        node_dicts = validated_data['data']
+        nodes = [Node.load(node_dict['_id']) for node_dict in node_dicts]
+
+        changes_flag = False
+        for node in nodes:
+            if not node:
+                raise exceptions.NotFound
```
Comment (mfraezz) [inline api/institutions/serializers.py]: Can you give this a descriptive error message? e.g. `"Node <id> not found"`

File: api/institutions/views.py [hunk 25]
```diff
@@ -233,5 +239,89 @@ class InstitutionRegistrationList(InstitutionNodeList):
...
+
+    def perform_destroy(self, instance):
+        data = self.request.data['data']
+        user = self.request.user
+        ids = [datum['id'] for datum in data]
+        nodes = []
+        for id_ in ids:
+            node = Node.load(id_)
+            if not node.has_permission(user, osf_permissions.ADMIN):
+                raise exceptions.PermissionDenied
```
Comment (mfraezz) [inline api/institutions/views.py]: Please give this a message that indicates failing node `_id`.

File: api/nodes/serializers.py [hunk 26]
```diff
@@ -485,58 +485,88 @@ def get_absolute_url(self, obj):
...
-            return node
-        node.remove_primary_institution(user)
+        add, remove = self.get_institutions_to_add_remove(
+            institutions=node.affiliated_institutions,
+            new_institutions=validated_data['data']
+        )
+
+        for inst in add:
+            if inst not in user.affiliated_institutions:
+                raise exceptions.PermissionDenied
```
Comment (mfraezz) [inline api/nodes/serializers.py]: Please give this an error message indicating which institution for which the user lacks permission.

File: api/nodes/serializers.py [hunk 27]
```diff
@@ -485,58 +485,88 @@ def get_absolute_url(self, obj):
...
+        add, remove = self.get_institutions_to_add_remove(
+            institutions=node.affiliated_institutions,
+            new_institutions=validated_data['data']
+        )
+        if not len(add):
+            raise RelationshipPostMakesNoChanges
+
+        for inst in add:
+            if inst not in user.affiliated_institutions:
+                raise exceptions.PermissionDenied
```
Comment (mfraezz) [inline api/nodes/serializers.py]: Please give this an error message indicating the institution for which the user lacks permission.

**Comment by mfraezz** [comment]
Nitpicky changes and follow-up issues aside, LGTM. Should be RTM after these are addressed. Functional testing looked good. 🌅

File: api/institutions/serializers.py (hunk 2 above)
Comment (hmoco) [inline api/institutions/serializers.py]: I see, will do so!

**Comment by hmoco** [comment]
Improvement tickets:
https://openscience.atlassian.net/browse/OSF-6382
https://openscience.atlassian.net/browse/OSF-6381

**Comment by hmoco** [comment]
@mfraezz responded to your comments!

**Comment by mfraezz** [comment]
LGTM. RTM when the build passes 🐲

**Comment by sloria** [comment]
Feature branch created (consider this merged): `feature/multi-institutions`. Any further development should happen on that branch.

File: scripts/migration/migrate_primary_institutions.py [hunk 28]
```diff
@@ -0,0 +1,53 @@
...
+        inst = node.primary_institution
+        if inst not in node.affiliated_institutions:
+            node.affiliated_institutions.append(inst)
+        node.primary_institution = None
+        node.save()
+    if dry_run:
+        raise RuntimeError('Dry run, transaction rolled back.')
+
+def main():
+    dry_run = 'dry' in sys.argv
```
Comment (sloria) [inline scripts/migration/migrate_primary_institutions.py]: Minor: Current convention is to use `--dry`. I'll change this.
